### PR TITLE
Fix issues detected by Secure Development Life check.  (SDL-7.0)

### PR DESCRIPTION
**Pull request by rajan-chari**
Fix code around SDL warnings:
1) cast to uint64_t when needed
2) handle Winsock errors
3) handle file rename() errors
4) handle realloc() failure
5) revert range based for loops to iterator based for loops

### CODE REVIEW

**Comment by JohnLangford** [comment]
There's a barf in the linux compile here: 
https://travis-ci.org/JohnLangford/vowpal_wabbit/builds/302803208#L1710

**Comment by rajan-chari** [comment]
Checked in my change.  Will it run the check again automagically or do I need to do something?

File: c_test/c_test.vcxproj [hunk 1]
```diff
@@ -217,4 +220,5 @@
     </PropertyGroup>
     <Error Condition="!Exists('$(SolutionDir)\.nuget\NuGet.targets')" Text="$([System.String]::Format('$(ErrorText)', '$(SolutionDir)\.nuget\NuGet.targets'))" />
   </Target>
+  <Import Project="..\sdl\SDL-7.0-NativeAnalysis.targets" />
```
Comment (JohnLangford) [inline c_test/c_test.vcxproj]: Do we want to do this for all windows builds?

File: test/RunTests [hunk 2]
```diff
@@ -821,12 +821,12 @@ exit $ErrorCount;
 __DATA__
 # Test 1:
 {VW} -k -l 20 --initial_t 128000 --power_t 1 -d train-sets/0001.dat \
-    -f models/0001.model -c --passes 8 --invariant \
+    -f models/0001_1.model -c --passes 8 --invariant \
```
Comment (JohnLangford) [inline test/RunTests]: Why the name change?

File: vowpalwabbit/array_parameters.h [hunk 3]
```diff
@@ -262,8 +262,8 @@ class sparse_parameters
 		// TODO: this is level-1 copy (weight* are stilled shared)
 		if (!_seeded)
 		{
-			for (auto& pair : _map)
-				free(pair.second);
+		  for (auto iter = _map.begin(); iter != _map.end(); ++iter)
```
Comment (JohnLangford) [inline vowpalwabbit/array_parameters.h]: Why use an explicit iterator?

File: vowpalwabbit/bs.cc [hunk 4]
```diff
@@ -41,9 +41,10 @@ void bs_predict_vote(example& ec, vector<double> &pred_vec)
   // float sum_labels = 0; // uncomment for: "avg on votes" and getLoss()
   bool majority_found = false;
   bool multivote_detected = false; // distinct(votes)>2: used to skip part of the algorithm
-  int* pred_vec_int = new int[pred_vec.size()];
+  auto pred_vec_sz = pred_vec.size();
+  int* pred_vec_int = new int[pred_vec_sz];
```
Comment (JohnLangford) [inline vowpalwabbit/bs.cc]: This isn't your fault, but generally speaking reductions should not use 'new' in the middle---we want to externalize memory allocations to reduce overhead.

File: c_test/c_test.vcxproj (hunk 1 above)
Comment (rajan-chari) [reply c_test/c_test.vcxproj]: Yes.  It degrades gracefully.  It will be ignored if SDL tools are not present.

File: test/RunTests (hunk 2 above)
Comment (rajan-chari) [reply test/RunTests]: I wanted to keep it consistent with other name changes.  I append "_<test#>" when I notice an issue.

**Comment by JohnLangford** [comment]
Look here: https://github.com/JohnLangford/vowpal_wabbit/pull/1360

-John

**Comment by arielf** [comment]
If these models aren't actually used, why store them in the 1st place.
I would recommend removing the `-f <modelname>` from these tests instead.

**Comment by JohnLangford** [comment]
Merged, thanks :-)  

Ariel, feel free to tweak tests.

**Pull request merged**